### ISSUE TITLE
type [1073741824]C.struct_Point larger than address space

### ISSUE THREAD

**Issue by stevef1uk** [hybridgroup/gocv#90]
Hi,

This is a great project! It works perfectly on my MacBook, but trying to use on a RPi3 I have built it locally on the RPi as documented successfully, but when trying to run the face detect example I see the above error at:

 go run main.go  0 ../../data/haarcascade_frontalface_default.xml 
# gocv.io/x/gocv
../../../../gocv.io/x/gocv/imgproc.go:57: type [1073741824]C.struct_Point larger than address space
../../../../gocv.io/x/gocv/imgproc.go:57: type [1073741824]C.struct_Point too large

Any thoughts?

Regards

Steve

**Comment by deadprogram** [comment]
Hello @stevef1uk which OS and version of Raspbian are you running? Also, which version of board. Lastly, have you tried to run any of the other examples on your setup?

**Comment by stevef1uk** [comment]
Hi,

I burnt a new rasbian version on a new SD card yesterday for this RPi3. 

Linux raspberrypi 4.9.59-v7+ #1047 SMP Sun Oct 29 12:19:23 GMT 2017 armv7l GNU/Linux

Don't know the board version, but I bought this one a couple of years ago and have just returned to it now I have taken a break from work.

Pi 3 Model B:
processor	: 0
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 38.40
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

...
I am currently using go go1.7.4 linux/arm

I am also trying to build go 1.10 on the RPi 3 but getting test failures!

Regards

Steve

**Comment by deadprogram** [comment]
I would recommend at last Go 1.8.x I have never tried any earlier versions with GoCV.

Also, try using a normal webcam. I know it is possible to use a raspicam but requires some additional setup. I have not had any time to work on that as yet.

**Comment by stevef1uk** [comment]
Ok. Thanks for the advice. I don't have a USB webcam I can try right now.

I just installed go1.10 from Google's distro and tried a different example: captest but get the same errors

**Comment by stevef1uk** [comment]
I 'bodged' it by adding two things.

1.  I am not sure whether this was needed but I changed the  Points structure to have __attribute__ ((aligned)); added in core.h

2. In improv.go line 57 I changed [1<<32] to [1<<20]

Then I was able to run go build in the main project directory ok.

However, when I try go test I get a lot of errors. The first part is:

go test
# _/home/pi/godev/src/github.com/gocv
/tmp/go-build053542233/b045/_x012.o: In function `cv::Mat::create(int, int, int)':
/usr/local/include/opencv2/core/mat.inl.hpp:784: undefined reference to `cv::Mat::create(int, int const*, int)'
/tmp/go-build053542233/b045/_x012.o: In function `cv::Mat::Mat(int, int, int, cv::Scalar_<double> const&)':
/usr/local/include/opencv2/core/mat.inl.hpp:421: undefined reference to `cv::Mat::operator=(cv::Scalar_<double> const&)'

<snip>

Is this a LD_LIBRARY_PATH issue or something more fundamental?

Regards

Steve

P.S. I found a CleverDog USB WebCam lying around home, but this is a wifi webcam and when connected to the RPi I can't seem to use it

**Comment by deadprogram** [comment]
I'm not sure about your changes, since those types are just wrappers about the underlying opencv types. Changing widths will for sure not work.

How exactly did you install OpenCV on the Raspi? Also, did you run the `env.sh` before running the Go program?

**Comment by stevef1uk** [comment]
Yes, you are right. I manually built OpenCV following various blogs on the internet. I only have 3.2 built so will soon try 3.4.

I had forgotten to run the env.sh script. 

I have found that on the RPi3 the maximum size before error is [1<<27) on that line in improv.go

The motion sensor example works really well on the RPi3 using the old raspicam. The facial recognition is pretty unreliable and very slow on the RPI3, but it isn't located in an ideal location and it is quite dark now. It was finding human faces in my light & curtain though and ignoring me :-)

Regards

Steve

**Comment by 0xIAx** [comment]
Hello,

I have the same problem with raspberry :/
Linux raspberrypi 4.9.59-v7+ #1047 SMP Sun Oct 29 12:19:23 GMT 2017 armv7l GNU/Linux

I treid with go 1.8 but have the same error.. in my case simplest version app doesn't work :(

pi@raspberrypi:~ $ cd /home/pi/go/src/gocv.io/x/gocv
pi@raspberrypi:~/go/src/gocv.io/x/gocv $ source ./env.shEnvironment variables configured for Linux
pi@raspberrypi:~/go/src/gocv.io/x/gocv $ go run ./cmd/version/main.go# gocv.io/x/gocv
./imgproc.go:57: type [1073741824]C.struct_Point larger than address space
./imgproc.go:57: type [1073741824]C.struct_Point too large
pi@raspberrypi:~/go/src/gocv.io/x/gocv $ 
pi@raspberrypi:~/go/src/gocv.io/x/gocv $ go version
go version go1.8 linux/arm
pi@raspberrypi:~/go/src/gocv.io/x/gocv $ 

Any Ideas? Or GoCV isn't a Pi friedly at all? :) Seems that problem not in a camera type itself.

Regards

Vadim

**Comment by stevef1uk** [comment]
I have it working on my RPi now. I made some notes to myself which may help you, sorry a bit random:

vcgencmd get_camera

(this checks that the PiCam is actually detected & usable)

Started from http://pklab.net/?id=392&lang=EN

Found I needed https://github.com/pixhawk/libdc1394 and and https://medium.com/@nszceta/fix-build-errors-on-python-3-5-when-compiling-opencv-3-fcf997cb6e

My successful build command:

cmake -D CMAKE_BUILD_TYPE=RELEASE -D CMAKE_INSTALL_PREFIX=/usr/local -D BUILD_WITH_DEBUG_INFO=OFF -D BUILD_DOCS=OFF -D BUILD_EXAMPLES=OFF -D BUILD_TESTS=OFF -D BUILD_opencv_ts=OFF -D BUILD_PERF_TESTS=OFF -D INSTALL_C_EXAMPLES=ON -D INSTALL_PYTHON_EXAMPLES=ON -D OPENCV_EXTRA_MODULES_PATH=../../opencv_contrib-3.4.0/modules -D ENABLE_NEON=ON  -D WITH_LIBV4L=ON ../

I also did:

sudo modprobe bcm2835-v4l2
export LD_LIBRARY_PATH=/usr/local/lib

The facial recognition demo finds faces in my wood walls & ceilings, which is funny. Obviously, it runs much faster on a mac.

**Comment by 0xIAx** [comment]
Thanks Steve,

I follow all steps but result is not success.. I see internal gcc error at the end of compilation opencv.. python libs as I can see. I will try to update to gcc 7.

Thanks for help!

**Comment by deadprogram** [comment]
I think it should be `INSTALL_PYTHON_EXAMPLES=OFF INSTALL_C_EXAMPLES=OFF` in this case, at least for using with GoCV.

**Comment by ZeleniJure** [comment]
I encountered (almost) the same problem.

1. took RASPBIAN STRETCH LITE (Release date:2018-03-13, kernel 4.9)
2. took [go v 1.10](https://dl.google.com/go/go1.10.1.linux-armv6l.tar.gz)
3. then did the following:

        cd ~
        go get -u -d gocv.io/x/gocv
        cd go/src/gocv.io/x/gocv
        make deps_debian  (-> make deps will detect fedora I think)
        make download
        make build

4. Environmental variables were not OK when sourcing from env.sh (there is no lib64 in my installation), so I did:

        source env.sh
        export LD_LIBRARY_PATH=/usr/local/lib
        export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig

5. then I ran `go run ./cmd/version/main.go` and got

        ./imgproc.go:39: type [1073741824]_Ctype_struct_Point larger than address space
        ./imgproc.go:39: type [1073741824]_Ctype_struct_Point too large
        ./imgproc.go:769: type [1073741824]_Ctype_struct_Point larger than address space
        ./imgproc.go:769: type [1073741824]_Ctype_struct_Point too large
        ./imgproc.go:1040: type [1073741824]_Ctype_struct_Point larger than address space
        ./imgproc.go:1040: type [1073741824]_Ctype_struct_Point too large

6. I fixed the `1 << 30` with `1 << 27` in imgproc and ran  `go run ./cmd/version/main.go`  again:

        gocv version: 0.11.0
        opencv lib version: 3.4.1

I didn't check code in detail, so I have no idea what the consequence of changing this is...

**Comment by iaburton** [comment]
Hi there!

I had been using this library on my PiZero W, using v0.4.2 and 'make install' to install opencv. The docs say if you already have opencv you don't need to reinstall it, but after updating this to v0.11 (an admittedly big jump) I was getting errors, so I reran 'make install' (an overnighter on a PiZero). After finishing I tried to run version/main.go and started getting the problems listed in this issue. Specifically about the struct pointer being too large.

It looks like the first commit to add one of these fields that are too large was here at line 57
https://github.com/hybridgroup/gocv/blob/4694c82862a9cb585f83a05bd3a26350bfa4149c/imgproc.go

Without more knowledge of how OpenCV works or is laid out I'm not sure on what the correct solution is either. Maybe those constants should be in architecture specific files? Would 32bit ARM version of OpenCV have the correct size somewhere we can use?

**Comment by suprememoocow** [comment]
I've submitted a PR for this https://github.com/hybridgroup/gocv/pull/194

**Comment by deadprogram** [comment]
This code has now been released as part of v0.12.0 so now closing. Thanks again!
